### PR TITLE
ath79: add support for Senao WatchGuard AP300 / AP200 / AP100

### PR DESCRIPTION
**Pull request by mcprat**
FCC ID: Q6G-AP300

WatchGuard AP300 is an indoor wireless access point with
1 Gb ethernet port, dual-band wireless,
internal antenna plates, and 802.3at PoE+

this board is a Senao device:
the hardware is equivalent to EnGenius EAP1750
the software is modified Senao SDK which is based on openwrt and uboot
including image checksum verification at boot time,
and a failsafe image that boots if checksum fails

**Specification:**

  - QCA9558 SOC		MIPS 74kc, 2.4 GHz WMAC, 3x3
  - QCA9880 WLAN	PCI card 168c:003c, 5 GHz, 3x3, 26dBm
  - AR8035-A PHY	RGMII GbE with PoE+ IN
  - 40 MHz clock
  - 32 MB FLASH		S25FL512S
  - 2x 64 MB RAM	NT5TU32M16
  - UART console	J10, populated
  - GPIO watchdog	GPIO 16, 20 sec toggle
  - 6 antennas		5 dBi, internal omni-directional plates
  - 5 LEDs		power, eth0 link/data, 2G, 5G
  - 1 button		reset

**MAC addresses:**

  MAC address labeled as ETH
  Only one Vendor MAC address in flash at art 0x0

  eth0 ETH  *:3c art 0x0
  phy1 ---- *:3d ---
  phy0 ---- *:3e ---

**Serial console access:**

  For this board, its not certain whether UART is possible
  it is likely that software is blocking console access

  the RX line on the board for UART is shorted to ground by resistor R176
  the resistors R175 and R176 are next to the UART RX pin at J10

  however console output is garbage even after this fix

**Installation:**

  Method 1: OEM webpage

    use OEM webpage for firmware upgrade to upload factory.bin

  Method 2: root shell access

    downgrade XTM firewall to v2.0.0.1
    downgrade AP300 firmware: v1.0.1
    remove / unpair AP from controller
    perform factory reset with reset button
    connect ethernet to a computer
    login to OEM webpage with default address / pass: wgwap
    enable SSHD in OEM webpage settings
    access root shell with SSH as user 'root'
    modify uboot environment to automatically try TFTP at boot time
    (see command below)

    rename initramfs-kernel.bin to test.bin
    load test.bin over TFTP (see TFTP recovery)
    (optionally backup all mtdblocks to have flash backup)
    perform a sysupgrade with sysupgrade.bin

  NOTE: DHCP is not enabled by default after flashing

**TFTP recovery:**

  server ip: 192.168.1.101

  reset button seems to do nothing at boot time...
  only possible with modified uboot environment,
  running this command in the root shell:

    `fw_setenv bootcmd 'if ping 192.168.1.101; then tftp 0x82000000 test.bin && bootm 0x82000000; else bootm 0x9f0a0000; fi'`

  and verify that it is correct with

    `fw_printenv`

  then, before boot, the device will attempt TFTP from 192.168.1.101
  looking for file 'test.bin'

  to return uboot environment to normal:

    `fw_setenv bootcmd 'bootm 0x9f0a0000'`

**Return to OEM:**

  user should make backup of MTD partitions
  and write the backups back to mtd devices
  in order to revert to OEM
  (see installation method 2)

  It may be possible to use sysupgrade
  with an OEM image as well...
  (not tested)

**OEM upgrade info:**

  The OEM upgrade script is at /etc/fwupgrade.sh

  OKLI kernel loader is required because the OEM software
  expects the kernel to be no greater than 1536k
  and the factory.bin upgrade procedure would otherwise
  overwrite part of the kernel when writing rootfs.

**Note on eth0 PLL-data:**

  The default Ethernet Configuration register values will not work
  because of the external AR8035 switch between
  the SOC and the ethernet port.

  For QCA955x series, the PLL registers for eth0 and eth1
  can be see in the DTSI as 0x28 and 0x48 respectively.
  Therefore the PLL registers can be read from uboot
  for each link speed after attempting tftpboot
  or another network action using that link speed
  with `md 0x18050028 1` and `md 0x18050048 1`.

  The clock delay required for RGMII can be applied
  at the PHY side, using the at803x driver `phy-mode`.
  Therefore the PLL registers for GMAC0
  do not need the bits for delay on the MAC side.
  This is possible due to fixes in at803x driver
  since Linux 5.1 and 5.3

**Note on WatchGuard Magic string:**

  The OEM upgrade script is a modified version of
  the generic Senao sysupgrade script
  which is used on EnGenius devices.

  On WatchGuard boards produced by Senao,
  images are verified using a md5sum checksum of
  the upgrade image concatenated with a magic string.
  this checksum is then appended to the end of the final image.

Tested-by: Alessandro Kornowski <ak@wski.org>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

### CODE REVIEW

**Comment by mpratt14** [comment]
rebased, and a small fix for the new variable

File: target/linux/ath79/dts/qca9558_watchguard_ap300.dts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+		wifi_green {
+			label = "green:wifi";
+			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+	};
+
+	virtual_flash {
+		devices = <&fwconcat0 &fwconcat1 &fwconcat2>;
```
Comment (CodeFetch) [inline target/linux/ath79/dts/qca9558_watchguard_ap300.dts:63]: I think there is a partition definition missing here and for fwconcat0 and fwconcat1, isn't it?

Comment (mpratt14) [reply target/linux/ath79/dts/qca9558_watchguard_ap300.dts:63]: it is using the partition definitions from the DTSI

the only difference is that there is an extra partition for firmware space, so `devices` is redefined to include fwconcat2

**Comment by mpratt14** [comment]
marking draft because I'm adding more boards to this PR

**Comment by adschm** [comment]
Ping. Status?

In any case, mtd-mac to nvmem conversion is required ...

**Comment by mpratt14** [comment]
@adschm I have become busy with other things, and before I reopen this and #4080 I want to add the similar models (like AP200 and AP100 and AN700) because I found someone to work with on them.

**Comment by adschm** [comment]
Then please close the PRs until you plan to continue with them.

**Comment by mpratt14** [comment]
@adschm now set this to open, ready for review, except I am still finishing commit message and small details

**Review by Tolcheen — APPROVED** [review]
Last image dump you make works good on ap200 and ap300

File: target/linux/ath79/image/common-senao.mk [hunk 2]
```diff
@@ -21,7 +21,13 @@ define Build/senao-tar-gz
 	$(CP) $@ $@.tmp/openwrt-$(word 1,$(1))-root.squashfs && \
 	$(TAR) -cp --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name \
 		$(if $(SOURCE_DATE_EPOCH),--mtime="@$(SOURCE_DATE_EPOCH)") \
-		-C $@.tmp . | gzip -9n > $@ && \
+		-C $@.tmp . | gzip -9n > $@
+
+	- [[ -n $(WATCHGUARD_MAGIC) ]] && [[ $@ == *factory.bin ]] && \
+	echo -n $(WATCHGUARD_MAGIC) | cat $@ - | md5sum | \
+		cut -d ' ' -f1 | tr -d '\n' > $@.md5 && \
+	cat $@.md5 >> $@
```
Comment (hauke) [inline target/linux/ath79/image/common-senao.mk:29]: Why don't you just define this as a extra step? Like this:
```
define Build/watchguard
   echo -n $(1) | cat $@ - | md5sum | cut -d ' ' -f1 | tr -d '\n' > $@.md5 && \
   cat $@.md5 >> $@ && \
   rm $@.md5
endef
```
And then call it like this:
```
define Device/watchguard
  $(Device/senao_loader_okli)
  DEVICE_VENDOR := WatchGuard
  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
	check-size | senao-tar-gz $$$$(SENAO_IMGNAME) | watchguard $$$$(WATCHGUARD_MAGIC)
endef
```
I haven't tested these examples and I assume they contain some errors.

Comment (mpratt14) [reply target/linux/ath79/image/common-senao.mk:29]: I agree that's much cleaner

originally I was avoiding to have to redefine factory.bin recipe, but I suppose it's worse to have the bash conditionals

**Comment by hauke** [comment]
The pull request is still marked as work in progress, is this correct?

**Comment by mpratt14** [comment]
it used to be WIP

currently this is ready except to update the merge conflict and the idea you just put

**Comment by mpratt14** [comment]
- updated per review from @hauke 
 - merge conflict fixed
 - added to commit messages

I am now build-testing one last time

also there is an empty line to clean up
and I want to get Tested-by line from those who have the device

**Comment by mpratt14** [comment]
@hauke @adschm ok I believe this is ready now. build test is good, the extra checksum is appended correctly

commit messages updated, including at least 1 tested-by each, but more can be tagged if they want

File: target/linux/ath79/generic/base-files/etc/board.d/01_leds [hunk 3]
```diff
@@ -40,8 +40,11 @@ alfa-network,n5q)
 	ucidef_set_led_rssi "signal3" "SIGNAL3" "green:signal3" "wlan0" "50" "100"
 	ucidef_set_led_rssi "signal4" "SIGNAL4" "green:signal4" "wlan0" "75" "100"
 	;;
-alfa-network,pi-wifi4)
-	ucidef_set_led_netdev "lan_data" "LAN_DATA" "orange:lan_data" "eth0" "tx rx"
+alfa-network,pi-wifi4|\
+watchguard,ap100|\
+watchguard,ap200|\
+watchguard,ap300)
+	ucidef_set_led_netdev "lan_data" "LAN_DATA" "orange:lan_data" "eth0" "rx"
```
Comment (hauke) [inline target/linux/ath79/generic/base-files/etc/board.d/01_leds:47]: Why do you change this from `"tx rx"` to `"rx"` ?

Comment (mpratt14) [reply target/linux/ath79/generic/base-files/etc/board.d/01_leds:47]: It's not very important, but...

if the other host is dead (powered link only), or if the ethernet is not configured properly (frames failing checksum) then "tx" blinking still occurs, but is not an indicator of real traffic making it from one side to another. it is only "rx" that tells you that real data is making it from the physical layer to the mac layer.

the amount of blinking is about the same in normal use with common traffic (but I suppose UDP-only traffic would not blink). it's useful in testing ethernet switch configuration.

if you think something like that should be changed in a separate commit, ill remove it here

**Comment by mpratt14** [comment]
@hauke I put the tx back. Can this merge before branching?

**Comment by Kyklas** [comment]
@mpratt14 Thanks for the image, I flashed an WG AP200. Works great.

**Comment by hauke** [comment]
Thank you for the patches, I applied them to master in 513f9855e9ce710c379cb3af4cd86b899f637c6e, c10750688366d14146cf6d2f0b73fd9480b29537, 146aaeafb7e364ca667f6ddf5c1857148e8bffb2, 9f6e24785439d7838fc7a338505cbfac9687a90d and 5df1b33298d1b35a47d24e9b4b941d2786633fb2.

**Pull request closed without merging**